### ISSUE TITLE
Element.elm error

### ISSUE THREAD

**Issue by BKSpurgeon** [brandly/elm-minesweeper#1]
Hi there Matt

Excellent work on your minesweeper game. I played it and it works beautifully. I want to get it working on my PC (elm 0.19.1) so I can try and grok elm but am having some problems here:

```
-- TYPE MISMATCH  /home/koshy/UbuntuHDD/Documents/Coding/ELM-projects/elm-minesweeper/src/Element.elm

The 1st argument to `el` is not what I expect:

16|         el ([ style css ] ++ attrs) children
                ^^^^^^^^^^^^^^^^^^^^^^
This argument is:

    List (String -> Attribute msg)

But `el` needs the 1st argument to be:

    List (Attribute msg)


-- TYPE MISMATCH  /home/koshy/UbuntuHDD/Documents/Coding/ELM-projects/elm-minesweeper/src/Element.elm

The 1st argument to `style` is not what I expect:

16|         el ([ style css ] ++ attrs) children
                        ^^^
This `css` value is a:

    List ( String, String )

But `style` needs the 1st argument to be:

    String

```
### Pointers?
Was wondering if you had any pointers? 

Here is the code if you wanted to replicate exactly what I have: https://github.com/BKSpurgeon/elm-minesweeper

### Game graphics

Also I am curious how you did the graphics for minesweeper?

chrs

Ben

**Comment by brandly** [comment]
hello! thanks for helping update this to 0.19.

i think i see what's going on. `styled` is currently defined like this:

```elm
styled : Element msg -> List ( String, String ) -> Element msg
styled el css =
    \attrs children ->
        el ([ style css ] ++ attrs) children
```

and then it's used like this:

```elm
insetDiv : Element msg
insetDiv =
    styled div
        [ ( "border", "2px solid #7b7b7b" )
        , ( "border-bottom-color", "#fff" )
        , ( "border-right-color", "#fff" )
        ]

-- ...

view : Model -> Html Msg
view model =
    insetDiv [ onClick DoSomething ] [ h1 [] [ text "hello" ] ]
```

the idea is that `styled` helps define `insetDiv` which can be used in place of `div`, wherever you need those styles. it's inspired by [Style Components](https://www.styled-components.com/) but a cheap hack by comparison.

---

in 0.19, `style` (from `Html.Attributes`, used inside `styled) change its type signature.
```elm
-- old
style : List ( String, String ) -> Attribute msg
-- new
style : String -> String -> Attribute msg
```

luckily, i did the same thing in another project and already updated it to 0.19! [take a look](https://github.com/brandly/elm-dr-mario/blob/051e6e72b25f2cb0797bf72dda6ba4a973c0fe05/src/Element.elm#L11-L14) 😄 

---

> Also I am curious how you did the graphics for minesweeper?

i found [this image](https://github.com/joelbyrd/external-resources/blob/master/images/minesweeper.png), but i don't remember how.

i then set styles like this:
```css
background-image: url(https://raw.githubusercontent.com/joelbyrd/external-resources/master/images/minesweeper.png);
background-position: 0px -39px;
width: 16px;
height: 16px;
```

i wrote [some logic around this idea](https://github.com/brandly/elm-minesweeper/blob/master/src/Bitmap.elm), so that i can ignore these details in `Main.elm`.

---

using inline styles and hotlinking to an image from another repo really is the best way to do things, but my focus here was learning Elm and getting more comfortable with functional programming. these decisions were good enough for my use, but the right way to handle things in a "serious" project. 

thanks for taking an interest in this repo!

**Comment by BKSpurgeon** [comment]
ok cheers, I seem to have got that bit working. just trying to debug through the rest atm.

So with the pictures, you are cutting up the image manually and then displaying it accordingly: am I understanding this correctly?

**Comment by brandly** [comment]
yeah, that's the idea. maybe you're familiar with a "sprite," but i manually figured out which portions of the image corresponds to different types of cells. `Bitmap.elm` holds that knowledge.

the other key is these two pieces:

```elm
insetDiv : Element msg
insetDiv =
    styled div
        [ ( "border", "2px solid #7b7b7b" )
        , ( "border-bottom-color", "#fff" )
        , ( "border-right-color", "#fff" )
        ]


raisedDiv : Element msg
raisedDiv =
    styled div
        [ ( "border", "2px solid #7b7b7b" )
        , ( "border-top-color", "#fff" )
        , ( "border-left-color", "#fff" )
        ]
```

they provide the frame of the window that characterizes this Windows XP style.

**Comment by brandly** [comment]
send me a pull request when you get 0.19 working. 

i've got mechanics in place to handle easy/medium/hard difficulty, but there's no UI for choosing which one. maybe i'll get around to adding that.

**Comment by BKSpurgeon** [comment]
chrs thx @brandly 

I'm very slowly trying to make sense of it all. Hopefully i can get it working on 0.19.1. we'll see how it goes.

how did you learn elm to develop a game of this level of complexity?

**Comment by brandly** [comment]
> Hopefully i can get it working on 0.19.1. we'll see how it goes.

look at [this commit](https://github.com/brandly/elm-flame-graph/commit/20308481f8894ad629efad81377ab1a4a590ad72) where i upgraded some other code to `0.19.1`. maybe it'll help.

> how did you learn elm to develop a game of this level of complexity?

i had already written [_a lot_ of code](https://mattbrandly.com/projects/) before coming to Elm. most of that code was JavaScript, so i was already very familiar with web browsers. i wasn't very familiar with type systems tho.

i read [the guide](https://guide.elm-lang.org/) and started working on [a game](https://github.com/brandly/elm-slime-volleyball). some people i know write Elm (and Haskell), so i could ask them questions and discuss things that were confusing me. join the [elm lang slack](elmlang.slack.com), and there's a `#beginners` channel where people are helpful. you could also [watch some videos](https://www.youtube.com/channel/UCOpGiN9AkczVjlpGDaBwQrQ/videos) to see what other people are doing with Elm.

ultimately, i got better at thinking in functional programming terms, leveraging the type system, and reading compiler errors. coming from languages like JavaScript, types initially felt like a hindrance, but once i learned how to really use them, they helped clarify my thinking. like anything, it just takes practice.

i'm most proud of [this Elm project](https://github.com/brandly/elm-dr-mario), if you'd like to check it out. i had written Elm for about a year before starting it. i didn't touch it for a while, but then i worked on it more recently after having written Elm for _another_ year. i've used it as a way to apply my understanding and experiment with ideas, but i had to write a lot of smaller projects before getting to this point.

tl;dr - join that Slack, make something small, and then make something a bit bigger 😄

**Comment by BKSpurgeon** [comment]
ok Matt that is very helpful.

and that project is very impressive. If i can get to that level, I would be well pleased.

hopefully i can get minesweeper working on 0.19. it feels beyond what i can handle, but i hope to try nonetheless.

**Comment by brandly** [comment]
thanks! 😊 

> hopefully i can get minesweeper working on 0.19

i hope you've see the [upgrade guide](https://github.com/elm/compiler/blob/master/docs/upgrade-instructions/0.19.0.md). there's also [this program](https://github.com/avh4/elm-upgrade), but i'm not sure what it'll do if you've already converted the `elm-package.json` to `elm.json` by hand.

if you get stuck, shoot me a message.

---

i see a lot of C# on your page. is that what you write at your day job?

**Comment by BKSpurgeon** [comment]
ok cheers Matt

I didn't know that existed, and was doing everything by hand!

rgds

Ben

**Comment by BKSpurgeon** [comment]
> i see a lot of C# on your page. is that what you write at your day job?

I use c# to support CAD related plugins. probably 30-40% of my time is here. I also do Ruby on Rails projects which makes up another other 40%.

And I dabble in elm to get used to front end projects. I don't think i will every use javascript and/or jquery if i can avoid it. 

Dr Mario is impressive, as is minesweeper. it's inspiring to see such work.

Here is an example of the latest: https://vimeo.com/manage/375346074/general  little tool I wrote.

**Comment by brandly** [comment]
(that link didn't work for me, but i think it's [this one](https://vimeo.com/375346074).)

very cool! i don't know much at all about CAD, but that seems like very interesting work.

javascript gets a bad rap, but i enjoy it. elm has been great and really introduced me to type systems. Dr Mario wouldn't be as solid as it is without using Elm, but for some things, i'd still rather use javascript.

thanks for the kind words! i make things like this just to learn, so it's cool that you find it interesting.

**Comment by BKSpurgeon** [comment]
```elm
styled : Element msg -> List ( String, String ) -> Element msg
styled el css =
    \attrs children ->
        el (List.map (\( k, v ) -> style k v) css ++ attrs) children
```

Am having trouble understanding this function. Which is located in the Element.elm file [here it is](https://github.com/brandly/elm-minesweeper/blob/159aff0c9c271f2b1b014a0371dfedfb5d26ba32/src/Element.elm#L11) (I have updated it). I'm not understanding  where are the `attrs` and `children` variables coming from?

Clarification here woudl be appreciated.

**Comment by brandly** [comment]
i'll go through that file and explain what's going on:

```elm
type alias Element msg =
    List (Attribute msg) -> List (Html msg) -> Html msg
```

this is an alias i'm defining, but this is the general type signature of any html element e.g. `Html.div`. so let's look at the type signature of `styled`

```elm
--       first param    second param               return value
styled : Element msg -> List ( String, String ) -> Element msg
styled el css =
    \attrs children ->
        el ([ style css ] ++ attrs) children
```
i put comments above to refer to the pieces of the type signature.

`first param` is a function because the `Element msg` type is aliased to a function. that function would be something like `Html.div`.

the `return value` is the same kind of function.  this allows us to write code like this, producing `insetDiv` that can be used in place of `div`:

```elm
insetDiv : Element msg
insetDiv =
    styled div
        [ ( "border", "2px solid #7b7b7b" )
        , ( "border-bottom-color", "#fff" )
        , ( "border-right-color", "#fff" )
        ]
```

so the body of `styled` is returning a function that takes `attrs` and `children`:

```elm
    \attrs children ->
        el ([ style css ] ++ attrs) children
```

that returned function can take arbitrary html attributes `attrs` and arbitrary html `children` to nest inside this `el`.

since this is the older version of elm, `style` takes a `List ( String, String )` and produces an `Attribute msg`. we prepend `attrs` with this extra attribute full of css.

so when we say `insetDiv [] [text "hi"]`, this expands out like so:

```elm
    -- this expands out
    -- el: div
    -- attrs: [] 
    --  children: [text "hi"]
    \attrs children ->
        el ([ style css ] ++ attrs) children

    -- like so
        div ([ style css ] ++ []) [text "hi"]
```

make more sense?

**Comment by BKSpurgeon** [comment]
I read it many, many times. 

I still do not quite get it.

```elm
--       first param    second param               return value
styled : Element msg -> List ( String, String ) -> Element msg
styled el css =       -- is 'el' the first param noted above, and is 'css' the second param noted above?
    \attrs children ->
        el ([ style css ] ++ attrs) children
```

**Comment by brandly** [comment]
yes!

```elm
el : Element msg
css : List ( String, String )

returnValue : Element msg
returnValue = 
    \attrs children ->
        el ([ style css ] ++ attrs) children
```

and remember this type alias. it's just another name for the same thing:

```elm
type alias Element msg =
    List (Attribute msg) -> List (Html msg) -> Html msg
```

without that alias, we can describe their types like this:

```elm
el : List (Attribute msg) -> List (Html msg) -> Html msg
css : List ( String, String )

returnValue : List (Attribute msg) -> List (Html msg) -> Html msg
returnValue = 
    \attrs children ->
        el ([ style css ] ++ attrs) children
```
that anonymous function we're calling `returnValue` is used to define views, exactly like you might use `Html.div`. here's that same example from an earlier comment fully turned into a `view`:

```elm
styled : Element msg -> List ( String, String ) -> Element msg
styled el css =
    \attrs children ->
        el ([ style css ] ++ attrs) children

-- passing these values into `styled`
-- el = div
-- css = [ ( "border", "2px solid #7b7b7b" ), ... ]

insetDiv : Element msg
insetDiv =
    styled div
        [ ( "border", "2px solid #7b7b7b" )
        , ( "border-bottom-color", "#fff" )
        , ( "border-right-color", "#fff" )
        ]

-- which returns this function to the variable `insetDiv`
--  \attrs children ->
--        div ([ style [ ( "border", "2px solid #7b7b7b" ), ... ] ] ++ attrs) children

-- we need a view...

type Msg = Something | Other

view : Html Msg
view = insetDiv [ onClick Something ] [ text "hello" ]

-- we pass values 
-- attrs = [ onClick Something ]
-- children = [ text "hello" ]

-- into `insetDiv` 
\attrs children ->
    div ([ style [ ( "border", "2px solid #7b7b7b" ), ... ] ] ++ attrs) children

-- which becomes 
div ([ style [ ( "border", "2px solid #7b7b7b" ), ... ] ] ++ [ onClick Something ]) [ text "hello" ]

-- we can resolve this further 
div [ style [ ( "border", "2px solid #7b7b7b" ), ... ] , onClick Something ] [ text "hello" ]
```

ultimately, this is nice because we can just write `insetDiv [ onClick Something ] [ text "hello" ]` instead of that whole mess, but we still get a nicely styled div.

in this example, it's a lot of overhead for little benefit. when your view gets more complicated, i like this because the structure of the html is clearer when all those styles aren't distracting me.

**Comment by BKSpurgeon** [comment]
ok i'm on the home stretch. hopefully not my bugs to clear!

**Comment by BKSpurgeon** [comment]
Matt

Here is a Xmas present for you: https://github.com/BKSpurgeon/elm-minesweeper - got it working on 0.19.1. I can make a pull request, if you create a new 0.19.1 branch, you can merge to that and retain the original version. Please do have a look at the changes I made: hopefully they have solved problems rather than introduced them!

Still, I hope to go through the code more thoroughly.

rgds
Ben

**Comment by brandly** [comment]
great! i made a `0.19.1` branch for you to submit a PR. it'll make its way to `master` pretty quickly tho. thanks for helping with this!

**Comment by BKSpurgeon** [comment]
Hi Matt,

Just to confirm I"m understanding this:

```elm
type alias Element msg =
    List (Attribute msg) -> List (Html msg) -> Html msg


--       first param    second param               return value
styled : Element msg -> List ( String, String ) -> Element msg
styled el css =
    \attrs children ->
        el ([ style css ] ++ attrs) children


-- Can the above be re-written like this:
--            first param          second param                return value (which is a function of type Element)
styled :   Element msg ->  List ( String, String ) -> (((List (Attribute msg) -> List (Html msg) -> Html msg)))

-- and consequently, the returned value is a function that takes in two parameters: 
-- first param of Element msg    second param of Element msg   (please confirm)
-- List (Attribute msg)          List (Html msg)
   \attrs                        children ->
                                              el ([ style css ] ++ attrs) children
```

If that's correct, I think I understand it, theoretically speaking.

chrs.

**Comment by brandly** [comment]
Sorry for not responding to that comment. Looks correct to me tho. Everything good? Everything making sense?

**Comment by BKSpurgeon** [comment]
Hi Matt, it's making sense, but only theoretically. The currying of the functions is what I had missed. They key point is that styled returns a function.
